hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReaderImpl.java

```
@@ -0,0 +1,284 @@
+    if (compareCurrent > 0) {
+      if (currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        int comparedNextBlockFirstKey =
+            key.compareTo(currentDataBlockEntry.get().getNextBlockKey().get());
+        if (comparedNextBlockFirstKey >= 0) {
+          // Searches the block that may contain the lookup key based the starting keys of
+          // all blocks (sorted in the TreeMap of block index entries), using binary search.
+          // The result contains the greatest key less than or equal to the given key,
+          // or null if there is no such key.
+
+          Map.Entry<Key, BlockIndexEntry> floorEntry = dataBlockIndexEntryMap.floorEntry(key);
+          if (floorEntry == null) {
+            // Key smaller than the start key of the first block
+            return SEEK_TO_BACKWARDS;
+          }
+          currentDataBlockEntry = Option.of(floorEntry.getValue());
+          currentDataBlock = Option.empty();
+          currentPos.setOffset(
+              (int) currentDataBlockEntry.get().getOffset() + HFILEBLOCK_HEADER_SIZE);
+        }
+      }
+      if (!currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        // This is the last data block.  Check against the last key.
+        if (fileInfo.getLastKey().isPresent()) {
+          int comparedLastKey = key.compareTo(fileInfo.getLastKey().get());
+          if (comparedLastKey > 0) {
+            currentDataBlockEntry = Option.empty();
+            currentDataBlock = Option.empty();
+            currentPos.setEof();
+            return SEEK_TO_EOF;
+          }
+        }
+      }
+
+      if (!currentDataBlock.isPresent()) {
+        currentDataBlock = Option.of(instantiateHFileDataBlock(currentDataBlockEntry.get()));
+      }
+
+      return currentDataBlock.get()
+          .seekTo(currentPos, key, (int) currentDataBlockEntry.get().getOffset());
+    }
+    if (compareCurrent == 0) {
+      return SEEK_TO_FOUND;
+    }
+    // Backward seek not supported
```

not supported? is that just a comment or should we throw an exception if unsupported

---

I've added the logic to detect backward seeks, separate from the keys less than the first key.  In that case, exception is thrown.